helix-gateway/src/main/java/org/apache/helix/gateway/channel/HelixGatewayServicePollModeChannel.java

```diff
@@ -20,38 +20,195 @@
+      if (clusterDiffMap == null || clusterDiffMap.isEmpty()) {
+        continue;
+      }
+      for (String instanceName : clusterDiffMap.keySet()) {
+        // if the instance is previously connected, send state change event
+        if (_livenessResults.get(clusterName) != null && _livenessResults.get(clusterName).get(instanceName)) {
+          logger.info("Host {} has state change, sending event to gateway manager", instanceName);
+          pushClientEventToGatewayManager(_manager,
+              StateTransitionMessageTranslateUtil.translateCurrentStateChangeToEvent(clusterName, instanceName,
+                  clusterDiffMap.get(instanceName)));
+        }
+      }
+      currentStateDiff.put(clusterName, clusterDiffMap);
+    }
+
+    // 2. fetch host health
+    for (String clusterName : _livenessCheckEndpointMap.keySet()) {
+      for (String instanceName : _livenessCheckEndpointMap.get(clusterName).keySet()) {
+        boolean prevLiveness =
+            _livenessResults.get(clusterName) != null && _livenessResults.get(clusterName).get(instanceName);
+        boolean liveness = fetchInstanceLivenessStatus(clusterName, instanceName);
+
+        if (prevLiveness && !liveness) {  // previously connected, now disconnected
+          logger.warn("Host {} is not healthy, sending event to gateway manager", instanceName);
+          pushClientEventToGatewayManager(_manager,
+              StateTransitionMessageTranslateUtil.translateClientCloseToEvent(clusterName, instanceName));
+        } else if (!prevLiveness && liveness) {  // new connection.
+          logger.info("Host {} is newly connected, sending init connection event to gateway manager", instanceName);
+          pushClientEventToGatewayManager(_manager,
+              StateTransitionMessageTranslateUtil.translateCurrentStateDiffToInitConnectEvent(clusterName, instanceName,
+                  currentStateDiff.containsKey(clusterName) ? currentStateDiff.get(clusterName).get(instanceName)
+                      : new HashMap<>()));
+        }
+        _livenessResults.computeIfAbsent(clusterName, k -> new HashMap<>()).put(instanceName, liveness);
+      }
+    }
   }
 
   @Override
   public void sendStateChangeRequests(String instanceName,
       HelixGatewayServiceOuterClass.ShardChangeRequests shardChangeRequests) {
+    switch (_shardStateChannelType) {
+      case FILE:
+        // we are periodically writing to the file, so no need to write here.
+        break;
+      default:
+        throw new NotImplementedException("Only support file based channel for now");
+    }
   }
 
   @Override
   public void start() throws IOException {
+    logger.info("Starting Helix Gateway Service Poll Mode Channel...");
+    final Runnable fetchUpdatesTask = new Runnable() {
+      @Override
+      public void run() {
+        fetchUpdates();
+      }
+    };
+    _scheduler.scheduleAtFixedRate(fetchUpdatesTask, _config.getPollStartDelaySec(),  // init delay
```

Comment:
This is single threaded thread pool. Would that be possible one task stuck and blocking the entire work? Also do we need dedup.

Reply:
If one task stuck, then the gateway service is not functioning - we are not getting any data, either user current state or connect state. 
If it happen, no need to worry about the either task is running or not.

Reply:
Need metrics here then.

Reply:
Yes. Will add all metrics in following pr.